test/with_dummyserver/test_connectionpool.py

```diff
@@ -1115,3 +1116,49 @@ def test_pool_size_redirect(self):
         ) as pool:
             pool.urlopen("GET", "/redirect", preload_content=False)
             assert pool.num_connections == 1
+
+
+class TestConnectionNotDisplacedByNoneOnEmptyPoolError(SocketDummyServerTestCase):
+    def test_no_displacement(self):
+        block_send_event = Event()
+        handler_ready_event = self.start_basic_handler(block_send=block_send_event)
+        handler_ready_event.wait()
+
+        def consume_connection(pool):
+            pool.request(
+                "GET", "/", timeout=LONG_TIMEOUT, pool_timeout=SHORT_TIMEOUT,
+            )
+
+        with HTTPConnectionPool(
+            self.host, self.port, maxsize=1, block=True, retries=False,
+        ) as pool:
+            consume_connection_thread = Thread(target=consume_connection, args=(pool,))
+            consume_connection_thread.start()
+            try:
+                # Wait for connection pool is empty
+                while pool.num_requests < 1:
+                    time.sleep(0)
+                with pytest.raises(EmptyPoolError):
+                    consume_connection(pool)
+            finally:
+                block_send_event.set()
+                consume_connection_thread.join()
+
+            assert pool.pool.get(True, SHORT_TIMEOUT) is not None
+
+    def test_return_conn_to_pool_when_get_conn_failed(self):
```

Comment:
Maybe I'm not understanding this test, why would we want to return a connection to the pool if it raises an exception in the constructor?

Reply:
If we don't return the connection to the pool, we will get EmptyPoolError on every next call of self._get_conn.